File: go/day06.go
package main

import (
	"strings"
)

type position struct {
	i int
	j int
}

type state struct {
	p position
	d int
}

func move(pos position, dir int) position {
	if dir == 0 {
		return position{pos.i - 1, pos.j}
	} else if dir == 1 {
		return position{pos.i, pos.j + 1}
	} else if dir == 2 {
		return position{pos.i + 1, pos.j}
	} else if dir == 3 {
		return position{pos.i, pos.j - 1}
	}
	return position{0, 0}
}

func Day06(input string) (int, int) {
	// 	input = `....#.....
	// .........#
	// ..........
	// ..#.......
	// .......#..
	// ..........
	// .#..^.....
	// ........#.
	// #.........
	// ......#...`

	var grid [][]int // stores obstacles
	var pos position
	var init position
	dir := 0 // 0: up, 1: right, 2: down, 3: left

	for i, line := range strings.Split(input, "\n") {
		row := make([]int, len(line))
		for j, char := range line {
			if char == '#' {
				row[j] = 1
			}
			if char == '^' {
				init = position{i, j}
			}
		}
		grid = append(grid, row)
	}

	pos = position{init.i, init.j}

	visited := make(map[position]bool)
	visited[pos] = true

	for {
		new := move(pos, dir)
		if new.i < len(grid) && new.i >= 0 && new.j < len(grid[0]) && new.j >= 0 {
			if grid[new.i][new.j] == 1 {
				dir = (dir + 1) % 4
			} else {
				pos = new
				visited[pos] = true
			}
		} else {
			break
		}
	}

	ans1 := len(visited)
	ans2 := 0

	for i := range grid {
		for j := range grid[i] {
			if grid[i][j] == 0 && (i != init.i || j != init.j) {
				grid[i][j] = 1

				pos = position{init.i, init.j}
				dir = 0
				visitedDir := make(map[state]bool)
				visitedDir[state{pos, dir}] = true
				isLoop := false
				for {
					new := move(pos, dir)
					if new.i < len(grid) && new.i >= 0 && new.j < len(grid[0]) && new.j >= 0 {
						if grid[new.i][new.j] == 1 {
							dir = (dir + 1) % 4
						} else {
							pos = new
						}
						if visitedDir[state{pos, dir}] {
							isLoop = true
							break
						}
						visitedDir[state{pos, dir}] = true
					} else {
						break
					}
				}
				if isLoop {
					ans2 += 1
				}

				grid[i][j] = 0
			}
		}
	}

	return ans1, ans2
}

File: rust/src/day06.rs
use std::collections::HashSet;

#[derive(Clone, Hash, PartialEq, Eq)]
struct Position {
    i: usize,
    j: usize,
}

#[derive(PartialEq, Eq, Hash, Copy, Clone)]
enum Direction {
    Up,
    Right,
    Down,
    Left,
}

struct Grid {
    grid: Vec<Vec<i32>>,
}

#[derive(PartialEq, Eq, Hash)]
struct MoveState {
    pos: Position,
    dir: Direction,
}

impl Position {
    fn mov(&mut self, dir: &Direction) {
        match dir {
            Direction::Up => self.i -= 1,
            Direction::Right => self.j += 1,
            Direction::Down => self.i += 1,
            Direction::Left => self.j -= 1,
        }
    }

    fn next(&self, dir: &Direction) -> Position {
        let mut next = self.clone();
        next.mov(dir);
        next
    }
}

impl Direction {
    fn rotate(&mut self) {
        *self = match self {
            Direction::Up => Direction::Right,
            Direction::Right => Direction::Down,
            Direction::Down => Direction::Left,
            Direction::Left => Direction::Up,
        }
    }
}

impl Grid {
    fn obstacle_at(&self, pos: &Position) -> bool {
        self.grid
            .get(pos.i)
            .and_then(|r| r.get(pos.j))
            .map_or(false, |&x| x == 1)
    }

    fn next_in_bounds(&self, pos: &Position, dir: &Direction) -> bool {
        match dir {
            Direction::Up => pos.i > 0,
            Direction::Right => pos.j < self.grid[0].len() - 1,
            Direction::Down => pos.i < self.grid.len() - 1,
            Direction::Left => pos.j > 0,
        }
    }
}

fn infinite_loop(grid: &Grid, init: Position) -> bool {
    let mut pos = init.clone();
    let mut dir = Direction::Up;

    let mut states = HashSet::new();

    loop {
        if grid.next_in_bounds(&pos, &dir) {
            if grid.obstacle_at(&pos.next(&dir)) {
                dir.rotate();
            } else {
                pos.mov(&dir);
            }
            let state = MoveState {
                pos: pos.clone(),
                dir: dir.clone(),
            };
            if states.contains(&state) {
                return true;
            }
            states.insert(state);
        } else {
            return false;
        }
    }
}

pub fn run(input: &str) -> (i32, i32) {
    let i = input
        .lines()
        .map(|line| line.find('^'))
        .position(|x| x.is_some())
        .unwrap();
    let j = input.lines().nth(i).unwrap().find('^').unwrap();

    let init: Position = Position { i, j };

    let mut grid = Grid {
        grid: input
            .lines()
            .map(|line| line.chars().map(|c| if c == '#' { 1 } else { 0 }).collect())
            .collect(),
    };

    let mut pos = init.clone();
    let mut dir = Direction::Up;

    let mut visited = HashSet::new();
    visited.insert(pos.clone());

    loop {
        if grid.next_in_bounds(&pos, &dir) {
            if grid.obstacle_at(&pos.next(&dir)) {
                dir.rotate();
            } else {
                pos.mov(&dir);
                visited.insert(pos.clone());
            }
        } else {
            break;
        }
    }

    let mut count = 0;

    for pos in visited.iter() {
        grid.grid[pos.i][pos.j] = 1;

        if infinite_loop(&grid, init.clone()) {
            count += 1;
        }

        grid.grid[pos.i][pos.j] = 0;
    }

    (visited.len() as i32, count)
}
